flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java

```diff
@@ -38,215 +37,264 @@
-    return ofBindingCollection(elements, SdkBindingData::ofDatetime, LiteralTypes.DATETIME);
+  public static SdkBindingData<Map<String, Double>> ofFloatMap(Map<String, Double> map) {
+    return createMap(
+        map,
+        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.FLOAT)),
+        (value) -> BindingData.ofScalar(Scalar.ofPrimitive(Primitive.ofFloatValue(value))));
   }
 
-  public static SdkBindingData ofDurationCollection(List<Duration> elements) {
-    return ofBindingCollection(elements, SdkBindingData::ofDuration, LiteralTypes.DURATION);
+  public static SdkBindingData<Map<String, Long>> ofIntegerMap(Map<String, Long> map) {
+    return createMap(
+        map,
+        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.INTEGER)),
+        (value) -> BindingData.ofScalar(Scalar.ofPrimitive(Primitive.ofIntegerValue(value))));
   }
 
-  private static <T> SdkBindingData ofBindingCollection(
-      List<T> elements, Function<T, SdkBindingData> f, LiteralType literalType) {
-    List<SdkBindingData> bindings = elements.stream().map(f).collect(toUnmodifiableList());
+  public static SdkBindingData<Map<String, Boolean>> ofBooleanMap(Map<String, Boolean> map) {
+    return createMap(
+        map,
+        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.BOOLEAN)),
+        (value) -> BindingData.ofScalar(Scalar.ofPrimitive(Primitive.ofBooleanValue(value))));
+  }
 
-    return ofBindingCollection(bindings, literalType);
+  public static SdkBindingData<Map<String, Duration>> ofDurationMap(Map<String, Duration> map) {
+    return createMap(
+        map,
+        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.DURATION)),
+        (value) -> BindingData.ofScalar(Scalar.ofPrimitive(Primitive.ofDuration(value))));
   }
 
-  private static SdkBindingData ofBindingCollection(
-      List<SdkBindingData> elements, LiteralType elementType) {
-    List<BindingData> elementsIdl = new ArrayList<>();
+  public static SdkBindingData<Map<String, Instant>> ofDatetimeMap(Map<String, Instant> map) {
+    return createMap(
+        map,
+        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.DATETIME)),
+        (value) -> BindingData.ofScalar(Scalar.ofPrimitive(Primitive.ofDatetime(value))));
+  }
 
-    for (SdkBindingData element : elements) {
-      if (!element.type().equals(elementType)) {
-        String message =
-            String.format("Type %s doesn't match expected type %s", element.type(), elementType);
+  public static <T> SdkBindingData<List<T>> ofBindingCollection(
+      LiteralType literalType, List<SdkBindingData<T>> elements) {
+    List<BindingData> bindings = elements.stream().map(SdkBindingData::idl).collect(toList());
+    BindingData bindingData = BindingData.ofCollection(bindings);
 
-        throw new IllegalArgumentException(message);
-      }
+    boolean hasPromise = bindings.stream().anyMatch(SdkBindingData::isAPromise);
```

Comment:
This method (and other public methods that receive a literal type and SdkBindingData or function to extract one) should check that all the elements are of this same type as the one passed as argument
```suggestion
    List<LiteralType> incompatibleTypes = elements.stream()
            .map(SdkBindingData::type)
            .distinct()
            .filter(type -> !type.equals(literalType))
            .collect(toList());
    if (!incompatibleTypes.isEmpty()) {
      throw new IllegalArgumentException(
              String.format("Type mismatch: expected all elements of type %s but found some elements of type: %s", literalType, incompatibleTypes)
      );
    }

    boolean hasPromise = bindings.stream().anyMatch(SdkBindingData::isAPromise);
```